### PR TITLE
Added missing popular optimizers

### LINKED ISSUE #7: Implement popular optimizers
We should have most or all of the widely used optimizers in the ML literature:

- [ ] Adadelta
- [ ] Adagrad
- [ ] Adam
- [ ] AdamW 
- [ ] SparseAdam
- [ ] Adamax
- [ ] ASGD
- [ ] LBFGS
- [ ] NAdam
- [ ] RAdam
- [ ] RMSprop
- [ ] Rprop
- [ ] SGD

Currently only Adam and SGD are implemented, and using design decision that we can revisit. Most of these are implemented in PyMC, taken from the Lasagne library. We could just reuse these instead of rewriting our own. The most important thing is that the optimizers are easily "extendable" to include things like gradient clipping #5 and learning rate scheduling #6

### PR DESCRIPTION
**Pull request by HangenYuu**
## Description

Added missing popular optimizers

## Related Issue

- [ ] Closes #7 
- [ ] Related to #7

### CODE REVIEW

**Comment by HangenYuu** [comment]
Hi @jessegrabowski, I would like to help with implementing this library. It looks so fun, like the time I first learnt about Deep Learning! I already checked other optimizers and added AdamW first to open this PRs. Will add other optimizers followingly.

**Comment by jessegrabowski** [comment]
Hey! This is super awesome. Thanks for the work!

Two small things:

1. This PR won't actually close #7, just one of the items on the list. If you look at that list, you can click on the little circle by the AdamW line to create a new issue. that you can just call "Implement AdamW", and write a one-sentence note. Then this PR can close that.
2. Since we're going to do a bunch of optimizers, we should write some formal tests for them. I was lazy when I did the initial ones, so let's nip that in the bud. I'm not sure how we should do the test. Some ideas:
    1. We would require torch for developers and test against that? 
    2. Or work to a really simple case by hand (i.e. 1 parameter, one layer, no activation) and make sure we get the right answer? 
    3. Or just do what we're doing in the notebook now: make a little model, run it on a dataset for like 1000 steps, and make sure the error is going down? 

I guess I like (3) the best, because it seems easy :D

**Comment by HangenYuu** [comment]
> 1. This PR won't actually close

Is this just me or I don't see any littcle circle?
images

image

**Comment by HangenYuu** [comment]
> 2\. Since we're going to do a bunch of optimizers, we should write some formal tests for them.

I am currently using 3 but it only makes sure that the optimizer at least has SGD functionality. To check behavior, I think we also go with 1 and test the result against pytorch implementation.

**Comment by jessegrabowski** [comment]
You have to mouse over it first:

image

Anyway it's not like some important feature, basically I'm just suggesting that we do one PR per optimizer, so you open an issue for AdamW then close it with this, then another one for Adadelta, and so on. It will just make the git history easier to read down the road.

For the testing, maybe we can compare against [optax](https://github.com/google-deepmind/optax) instead? We already have a jax dependency, so that might be easier than torch.

**Comment by HangenYuu** [comment]
For the only scope of implementing AdamW, this PR can be merged and closed now.
